server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java

```diff
@@ -0,0 +1,189 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.geo;
+
+import org.apache.lucene.util.BitUtil;
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.geo.GeoUtils.normalizeLat;
+import static org.elasticsearch.common.geo.GeoUtils.normalizeLon;
+
+/**
+ * Implements quad key hashing, same as used by map tiles.
+ * The string key is formatted as  "zoom/x/y"
+ * The hash value (long) contains all three of those values.
+ */
+public class QuadkeyUtils {
+
+    /**
+     * Largest number of tiles (precision) to use.
+     * This value cannot be more than (64-5)/2 = 29, because 5 bits are used for zoom level itself
+     * If zoom is not stored inside hash, it would be possible to use up to 32.
+     * Another consideration is that index optimizes lat/lng storage, loosing some precision.
+     * E.g. hash lng=140.74779717298918D lat=45.61884022447444D == "18/233561/93659", but shown as "18/233561/93658"
+     */
+    public static final int MAX_ZOOM = 29;
+
+    /**
+     * Bit position of the zoom value within hash.  Must be &gt;= 2*MAX_ZOOM
+     * Keeping it at a constant place allows MAX_ZOOM to be increased
+     * without breaking serialization binary compatibility
+     * (still, the newer version should not use higher MAX_ZOOM in the mixed cases)
+     */
+    private static final int ZOOM_SHIFT = 29 * 2;
+
+    /**
+     * Mask of all the bits used by the quadkey in a hash
+     */
+    private static final long QUADKEY_MASK = (1L << ZOOM_SHIFT) - 1;
+
+    /**
+     * Parse quadkey hash as zoom, x, y integers.
+     */
+    private static int[] parseHash(final long hash) {
+        final int zoom = checkPrecisionRange((int) (hash >>> ZOOM_SHIFT));
+        final int tiles = 1 << zoom;
+
+        // decode the quadkey bits as interleaved xTile and yTile
+        long val = hash & QUADKEY_MASK;
+        int xTile = (int) BitUtil.deinterleave(val);
+        int yTile = (int) BitUtil.deinterleave(val >>> 1);
+        if (xTile < 0 || yTile < 0 || xTile >= tiles || yTile >= tiles) {
+            throw new IllegalArgumentException("hash-tile");
+        }
+
+        return new int[]{zoom, xTile, yTile};
+    }
+
+    private static double tile2lon(final double x, final double tiles) {
+        return x / tiles * 360.0 - 180;
```

Comment:
Tiny nit, should we put parenthesis around the components here just so the order of operation is immediately clear?

And not knowing anything about the algo, is there a chance `tiles` could ever be zero (e.g. divide by zero scenario)?

Reply:
I refactored to put all logic in a single method. Tiles is a result of (4^N), so cannot be 0